### ISSUE TITLE
K8S client issues

### ISSUE THREAD

**Issue by alacuku** [falcosecurity/libs#987]
**Motivation**

Falco has a built-in functionality called **Kubernetes Metadata Enrichment**. It provides k8s metadata, fetched from the k8s **api-server**, used by Falco to enrich the **system-call events**. Furthermore, these metadata are available to users as **[events fields](https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s)** to be used in the `conditions` and 'outputs' of Falco rules.

## Current implementation

The built-in k8s client is written in C++ and lives in the [libs repo](https://github.com/falcosecurity/libs). It has a number of issues discussed in the next sections.

### Poor performances
The **metadata** are fetched asynchronously from the **api-server** by opening a connection to the `watch` endpoint. Periodically, it checks for new events and processes them. The processing is done in the main loop, bloking the processing of *syscall* stream coming from the driver/probe. In large environments this causes syscall events to be dropped. For more info, please see: https://github.com/falcosecurity/falco/issues/2129. The following scheme shows how the client handles events coming from the **api-server** (some calls have been ommitted for the sake of clarity):

![built-in-k8s-client-flow](https://user-images.githubusercontent.com/17413176/223686298-6b52e081-b932-468e-a474-e3e45ec692c8.svg)

### Stability issues
The network connections to the **api-server** are far from being stable. It does not check for temporary network failures causing Falco to crash and restart. When deploying Falco in large clusters, all the Falco instances( one per node) connects to the **api-server**, and it could happen that some of them are **throtled** and the connection is closed. In such cases Falco pods will crash and restart. For more info see: https://github.com/falcosecurity/falco/issues/1909

### Scalability issues
A **watch** is opened toward the **api-server**for each resource. Currently, we collect metadata for the following resources:
* pods;
* deployments;
* replictioncontrollers;
* replicasets;
* daemonsets;
* namespaces;
* services;
* nodes.

In a 5K nodes cluster the Falco instances will open 40000 watches. This approach does not scale well and contributes to control plane outages in large clusters. 

Only Pods objects are efficiently retrieved by using the NodeSelector. For the other resource types, there is no way to filter them out, but we cache them locally leading to a significant waste of resources, especially memory.

**Feature**
## Native libs (client-go) based implementation

The idea is to introduce a centralized component that retrieves the K8s metadata required by Falco instances. The new component should be based on modern libraries such as the one provided by the k8s project. 

### Motivation

See the section about current implementation.

### Goals

* Fetch metadata from the k8s `api-server` using a centralized component;
* Keep the latency of the updates as short as possible;
* Send only the required metadata to Falco instances ready to be used without the need for further processing on Falco's side;

### Non-Goals

* Sharding the K8s state, the instance of the component will be responsible for the entire cluster;
* Adaptive metadata fetching based on requests by the Falco instances. We could think about it in the next iterations.

### Proposal

Provide a new component to be deployed in the K8s cluster responsible for providing the `metadata` required by each Falco instance. `Metadata` to be sent to Falco:

* Pod's metadata for a given node;
* Namespace metadata only for the ones where local pods reside;
*  Workloads metadata: Deployment, replicasets, daemonsets, replicationcontrollers;
* Services metadata only for the ones serving the local pods;

Each Falco instance will subscribe to the `centralized component` providing the `node name` where they reside and will receive metadata through a stream of updates. The communication between the Falco instances and the `centralized component` uses GRPC over HTTP2.

![Image](https://user-images.githubusercontent.com/17413176/226294705-2553c173-7c3d-4f85-bbe5-931e567e97c9.png)

Design Details

The **Meta Collector** will be implemented using Go relying on [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) to efficiently retrieve and cache K8s resources. An `operator/controller` will be implemented for each resource needed by Falco. Each operator will connect to the `api-server` and fetch the initial state for the watched resource. It will create a state in memory and send updates to the Falco instances that need the data. Since we are limited to a subset of the fields of the watched resources update events will be sent only when those fields change. Updates should be sent to Falco as soon as possible. Each operator will:
* Check if a `Modified` event received from the `api-server` produces an `update` to be sent to the Falco instances;
* Generate the event to be sent;
* Send the event to the `message broker` which will send it to the relevant Falco instances.

![Image](https://user-images.githubusercontent.com/17413176/226299056-7b8dcdc2-b5b1-429a-a0f0-86c8217b1816.png)

After the `operators` complete the initialization a GRPC server (message broker) will start and wait for Falco subscriptions. The message broker will compute an initial list of `events/updates` to send to the Falco instance upon a new subscription. Remember that the state is cached, so subscriptions will not generate requests to the `api-server`. Then the `updates` will be sent to the Falco instances as they come from the `api-server`.

**NOTE** This issue will be updated with more implementation details in the upcoming weeks.

**Comment by Andreagit97** [comment]
Thank you very much for this! We really need it as soon as possible for this reason I will put a milestone /0.11.0!  This is huge so of course, in case of blockers we can move it to the next milestone!

/milestone 0.11.0

BTW thank you very much for taking care of this!

**Comment by Dentrax** [comment]
What a well-written proposal. So kudos! It would definitely resolve our performance issues. Strong +1 from me.

**Comment by leogr** [comment]
I think we are too late to have this in Falco 0.35, so
/milestone 0.13.0

**Comment by leogr** [comment]
Cross-linking https://github.com/falcosecurity/plugins/pull/378

**Comment by Andreagit97** [comment]
yep to be honest I would move it to Falco because we will address it in Falco 0.37.0 rather than in libs 0.14.0
